### PR TITLE
Implements FSharpMemberOrFunctionOrValue.IsConstructorThisValue, IsMemberThisValue & LiteralValue

### PR DESCRIPTION
**Pull request by alfonsogarciacaro**
I didn't implement IsBaseValue, because I couldn't find a case where you can get a `base` reference without using `BasicPatterns.BaseValue` active pattern (which doesn't include a `FSharpMemberOrFunctionOrValue` value).

Also adds some tests for these properties and for the `FSharpMemberOrFunctionOrValue.Overloads`. The latter doesn't pass at the moment, see #356.

Can you please review the PR? Cheers!

### CODE REVIEW

File: src/fsharp/vs/Symbols.fs [hunk 1]
```diff
@@ -1491,18 +1492,33 @@ and FSharpMemberOrFunctionOrValue(cenv, d:FSharpMemberOrValData, item) =
...
     /// Is this the "x" in "type C() as x = ..."
-    member __.IsConstructorThisValue : bool
-
-    /// Is this the "x" in "member __.M = ..."
-    member __.IsMemberThisValue : bool
+    member __.IsConstructorThisValue =
+        if isUnresolved() then false else
+        match d with
+        | M _ | P _ | E _ -> false
+        | V v -> match v.BaseOrThisInfo with CtorThisVal -> true | _ -> false
```
Comment (dungpa) [inline src/fsharp/vs/Symbols.fs]: Could be shorted to `v.BaseOrThisInfo = CtorThisVal`.

File: src/fsharp/vs/Symbols.fs [hunk 2]
```diff
@@ -1491,18 +1492,33 @@ and FSharpMemberOrFunctionOrValue(cenv, d:FSharpMemberOrValData, item) =
...
 
-    /// Is this a [<Literal>] value, and if so what value?
-    member __.LiteralValue : obj // may be null
+    /// Is this the "x" in "member x.M = ..."
+    member __.IsMemberThisValue =
+        if isUnresolved() then false else
+        match d with
+        | M _ | P _ | E _ -> false
+        | V v -> match v.BaseOrThisInfo with MemberThisVal -> true | _ -> false
```
Comment (dungpa) [inline src/fsharp/vs/Symbols.fs]: Could be shorted to `v.BaseOrThisInfo = MemberThisVal `.

File: tests/service/ProjectAnalysisTests.fs [hunk 3]
```diff
@@ -4408,3 +4411,75 @@ let ``Test project35 CurriedParameterGroups should be available for nested funct
...
+    let options =  keepAssemblyContentsChecker.GetProjectOptionsFromCommandLineArgs (projFileName, args)
+    let wholeProjectResults =
+        keepAssemblyContentsChecker.ParseAndCheckProject(options)
+        |> Async.RunSynchronously
+    let declarations =
+        let checkedFile = wholeProjectResults.AssemblyContents.ImplementationFiles.[0]
+        match checkedFile.Declarations.[0] with
+        | FSharpImplementationFileDeclaration.Entity (_, subDecls) -> subDecls
+        | _ -> failwith "unexpected"
+    let getExpr exprIndex =
```
Comment (dungpa) [inline tests/service/ProjectAnalysisTests.fs]: I'm curious why you don't get all symbol uses and check their properties as per https://github.com/alfonsogarciacaro/FSharp.Compiler.Service/blob/dcfd9f816d807ff68fd120f6c406f65193b69f58/tests/service/ProjectAnalysisTests.fs#L4363-L4365. It seems complicated to walk the typed trees to assert simple properties.

File: tests/service/ProjectAnalysisTests.fs [hunk 4]
```diff
@@ -4408,3 +4411,75 @@ let ``Test project35 CurriedParameterGroups should be available for nested funct
...
+    match Project36.getExpr 3 with
+    | BasicPatterns.Call(_,_,_,_,[BasicPatterns.Value s;_])
+        when not s.IsMemberThisValue && not s.IsConstructorThisValue -> ()
+    | _ -> failwith "val s in B.Overload(s) must not be MemberThis"
+
+    let project36Module = Project36.wholeProjectResults.AssemblySignature.Entities.[0]
+    let lit = project36Module.MembersFunctionsAndValues.[0]
+    let notLit = project36Module.MembersFunctionsAndValues.[1]
+    if lit.LiteralValue.IsNone || notLit.LiteralValue.IsSome then
+        failwith "val lit must be LiteralValue while notLit musn't"
```
Comment (dungpa) [inline tests/service/ProjectAnalysisTests.fs]: Please try to use NUnit assertions (e.g. shouldEqual, etc.) rather than raising ad-hoc exceptions.

File: tests/service/ProjectAnalysisTests.fs [hunk 5]
```diff
@@ -4408,3 +4411,75 @@ let ``Test project35 CurriedParameterGroups should be available for nested funct
...
+
+[<Test>]
+let ``Test project36 FSharpMemberOrFunctionOrValue.Overloads(false)`` () =
```
Comment (dungpa) [inline tests/service/ProjectAnalysisTests.fs]: Do you plan to fix the bug related to `Overloads` in this PR? If not, this test doesn't belong to this PR.

**Comment by dungpa** [comment]
Look good. It would be great to fix the failing test if it's meant to be a part of the changes.

**Comment by alfonsogarciacaro** [comment]
Thanks a lot for your useful comments, @dungpa!

I've fixed the PR. You're right about the overloads (I had a look at the issue, but unfortunately my limited knowledge of the project didn't get me too far), so I deleted the test.

About walking the typed tree in the tests, I'm using the compiler services for FunScript where I'm accessing the symbols mostly through the AST and, as I've found some disagreements when you retrieve a symbol via the AST or something like `GetAllUsesOfAllSymbols` (as with `Overloads()`), I want to be sure the properties also work properly in the former case. I hope this is OK, but if it's not, tell me and I'll change the unit tests.

Cheers!

**Comment by dungpa** [comment]
Great jobs. Thanks a lot.

I'm not aware that anyone uses typed tree API extensively. There are probably inconsistencies to iron out.
